### DISCUSSION TITLE
How to debug CLI

### DISCUSSION THREAD

**Discussion by RunDevelopment** [wasm-bindgen/wasm-bindgen#4239] [Q&A]
### Summary

How I can attach a debugger to the CLI for tests. I want to debug the JS code gen part but couldn't find any docs for debugging.

I'm using VSCode btw and I cannot figure it out.

### Additional Details

I would also like to add how to do it to the docs once I know.

**Comment by daxpedda** [comment]
Is there anything special about `wasm-bindgen` that prevents regular debugging from working?

A quick google just gave me this: https://code.visualstudio.com/docs/languages/rust#_debugging, which is pretty much what I use as well.

**Reply by RunDevelopment** [reply]
I realize now how badly I worded this. I am mostly running `references.rs` for tests, and I cannot debug the CLI from there. 

Debugging doesn't need to work with `references.rs` per se. I just need some setup that lets me define a `.rs` file with `#[wasm_bindgen]` attrs and generates the JS output from it with a debugger. My breakpoints are in `bindings.rs` and the code around it, and nothing hits those breakpoints.

**Reply by daxpedda** [reply]
Can I assume that you got it working with #4242?

**Reply by RunDevelopment** [reply]
No. https://github.com/rustwasm/wasm-bindgen/pull/4242 just made debug work with the code in `reference.rs`. I still haven't been able to debug the actual JS code gen.

**Reply by daxpedda** [reply]
Probably the easiest way I know of to do what you want here is to just start the CLI binary through the debugger directly. 
Ergo, just click "Debug" on `fn main()` in `crates/cli/wasm-bindgen.rs` and setup the right [arguments in `launch.json`](https://code.visualstudio.com/docs/editor/debugging#_launchjson-attributes).

Alternatively you could just call the `wasm-bindgen-cli-support` library directly, then the debugger should work.

AFAIK VSCode doesn't possess any auto attach capability for debugging, but I found [this guide for LLDB extension](https://github.com/vadimcn/codelldb/blob/v1.11.1/MANUAL.md#attaching-debugger-to-the-current-process-rust), this way you should be able to *tell* VSCode to attach the debugger directly from your code!

**Reply by daxpedda** [reply]
I simply added [the lines of code as suggested in the guide](https://github.com/vadimcn/codelldb/blob/v1.11.1/MANUAL.md#attaching-debugger-to-the-current-process-rust) into `fn main()` in `crates/cli/wasm-bindgen.rs` and it actually worked like a charm.

So this might be the easiest way to do it.

**Reply by RunDevelopment** [reply]
Thanks for the guides! I'll try them tomorrow.

On a different note: I just had a thought. Why is `reference.rs` calling the CLI via `Command::cargo_bin`? Sure, E2E is nice, but couldn't we use the `Bindgen` API directly to bypass the CLI being separate binary? This would make debugging directly with `reference.rs` possible.

**Reply by daxpedda** [reply]
Because of that E2E you mentioned!
